[clangd] Re-land "support outgoing calls in call hierarchy"

HighCommander4 opened the pull request:
Previously reviewed at https://github.com/llvm/llvm-project/pull/77556 but was backed out due to build failures.

qchateau committed:
[clangd] Support outgoing calls in call hierarchy

The implementation is very close the the incoming
calls implementation. The results of the outgoing
calls are expected to be the exact symmetry of the
incoming calls.

Differential Revision: https://reviews.llvm.org/D93829

HighCommander4 committed:
Implement the RefKind::Call optimization

HighCommander4 committed:
Implement the simple lookup optimization

HighCommander4 committed:
Address other review comments

HighCommander4 committed:
Address newer review comments

HighCommander4 commented:
I've fixed the build failure that was reported by the majority of the buildbots (which is just a new test case using a test utility function that the patch renames).

However, there is an additional build failured reported by [this buildbot](https://github.com/llvm/llvm-project/pull/77556#issuecomment-2499665944) which only applies to builds with `-DCLANGD_ENABLE_REMOTE=ON`. There is a `SymbolIndex` implementation which is only compiled in that configuration (`clang::clangd::remote::IndexClient`), which needs to be implement the new virtual method `containedRefs()` introduced by the patch.

kadircet reviewed (changes requested):
sorry for losing track on the original patch, and thanks a lot for driving this to completion!

> @kadircet perhaps you might be able to pick up this review?
>
> Or, in the absence of a full review, your opinion on the directional question in [this comment](https://reviews.llvm.org/D93829#4654786) would be appreciated as well:
>
> > how would you feel about proceeding with the patch in its current state, with the memory usage increase brought down from 8.2% to 2.5% thanks to the combination of the simple lookup optimization + RefKind filtering, and leaving the "deep lookup optimization" to be explored in a future change?

I'd definitely prefer the one we discussed in the original review, but I don't think perfect needs to be enemy of the good, we can surely optimize data structures here going forward if needed.

I am still worried about certain deployments (chromium remote-index service, and some internal ones we have). Since the optimization relies on filtering by kind, memory usage increase might be more severe for projects with different structure. 

So if it isn't going to be too annoying to ask at this point, can we make all of this a config knob? Both handling of `callHierarchy/outgoingCalls` and population of relevant structures in `DexIndex`. Option can be turned on by default, I'd like to make sure we have a quick way out if some of those deployments start OOM'ing.

pidgeon777 commented:
I can't find the watch button for this pull request, so I'm just commenting: hopefully this will finally be merged 🎉.

HighCommander4 committed:
Implement remote index support for the new containedRefs request

HighCommander4 committed:
Bump index version to start storing RefKind::Call

HighCommander4 commented:
@kadircet Thanks for having a look. I made a few updates to the PR:

 * Implemented remote index support (https://github.com/llvm/llvm-project/pull/117673/commits/d380984cecf18b53b6e2fdd4266702dcb8ea513a)
 * Implemented a config option (https://github.com/llvm/llvm-project/pull/117673/commits/02bd8f6e342c7dfbedd18c791c9e73e3c0ccb385). I did run into some challenges/limitations here:
   * Only a user config option is respected, not a project config option. I think this limitation is necessary, because the index is a global structure rather than a per-project structure. (Concretely, in the call to the `Dex` constructor that happens via `BackgroundIndexRebuilder::maybeRebuild()`, what path would I pass to the context provider that gets it to look at an appropriate project config file?)
   * Only the `Dex` logic is conditioned on the option, not the binding of the `outgoingCalls` message. This is because the binding of the message happens fairly early, while we're processing the `initialize` request. At this point we haven't loaded config files yet, and if we try, that triggers a `publishDiagnostics` message for the config file, which confuses clients who are expecting an `initialize` response.
 * Bumped index version (https://github.com/llvm/llvm-project/pull/117673/commits/c493674c8519133e1e202194fa55564e75c5a08c)

HighCommander4 commented on clang-tools-extra/clangd/ClangdServer.cpp:
```diff
@@ -88,9 +90,12 @@ struct UpdateIndexCallbacks : public ParsingCallbacks {
       trace::Span Tracer("PreambleIndexing");
+      std::optional<WithContext> WithProvidedContext;
+      if (ContextProvider)
+        WithProvidedContext.emplace(ContextProvider(""));
```
@kadircet Not sure if this is important: it disables the building of the `RevRefs` data structure for the dynamic index.

While we do have a `Path` available to pass to the context provider here, I'm not passing it to remain consistent with the background index codepath, which only looks at the user config as explained in my previous comment.

(Also, is it fine to capture `this` in this lambda, lifetime-wise?)

kadircet reviewed (changes requested):
I am sorry for my poor choice of words around `config knob`, I was trying to imply a knob like https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/clangd/ClangdServer.h#L111. I don't think the knob needs to be exposed to end-users directly, only having programmatic access is fine initially. if you choose to expose it right away, i think a command line flag is the most natural way, rather than a config field. 

I'd rather plumb it explicitly through the interfaces of file-index/file-symbol/dex (and others), i think using `context` here is somewhat abusing the functionality. if you feel like that's too much (especially considering the size of the patch), feel free to leave a FIXME at least, to plumb it properly later on.

> Only a user config option is respected, not a project config option. I think this limitation is necessary, because the index is a global structure rather than a per-project structure. (Concretely, in the call to the Dex constructor that happens via BackgroundIndexRebuilder::maybeRebuild(), what path would I pass to the context provider that gets it to look at an appropriate project config file?)

you're absolutely right here. that's the main reason i don't think config mechanism actually works here. they're mostly meaningful for per-project/file configurable features. not for global ones like index.

> Only the Dex logic is conditioned on the option, not the binding of the outgoingCalls message. This is because the binding of the message happens fairly early, while we're processing the initialize request. At this point we haven't loaded config files yet, and if we try, that triggers a publishDiagnostics message for the config file, which confuses clients who are expecting an initialize response.

again this is resulting from my poor choice of words, if we treat this as part of option struct, we'll have the value at startup time and can decide what to do.

HighCommander4 committed:
Add a ClangdServer option to allow disabling outgoing calls

HighCommander4 commented on clang-tools-extra/clangd/index/dex/Dex.cpp:
```diff
@@ -149,15 +150,18 @@ void Dex::buildIndex() {
-    return A.ref().Container < B.ref().Container;
-  });
+  if (Config::current().CallHierarchy.OutgoingCalls) {
+    vlog("WALDOWALDO Building RevRefs");
```
Oops, debug statement left in accidentally :laughing:

HighCommander4 commented:
> I was trying to imply a knob like https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/clangd/ClangdServer.h#L111. I don't think the knob needs to be exposed to end-users directly, only having programmatic access is fine initially.

Thanks, that is indeed quite a bit simpler.

I reworked the flag to be just a ClangdServer option, and plumbed it in explicitly, in https://github.com/llvm/llvm-project/pull/117673/commits/0aee7953ef1285bb327f20a01389e554fb44ffa8.

kadircet commented on clang-tools-extra/clangd/index/FileIndex.h:
```diff
@@ -69,7 +69,7 @@ enum class DuplicateHandling {
 class FileSymbols {
 public:
-  FileSymbols(IndexContents IdxContents);
+  FileSymbols(IndexContents IdxContents, bool SupportContainedRefs = true);
```
can we avoid the default param here?

kadircet commented on clang-tools-extra/clangd/index/Serialization.h:
```diff
@@ -83,7 +83,8 @@ std::string toYAML(const Ref &);
 std::unique_ptr<SymbolIndex> loadIndex(llvm::StringRef Filename,
-                                       SymbolOrigin Origin, bool UseDex = true);
+                                       SymbolOrigin Origin, bool UseDex = true,
+                                       bool SupportContainedRefs = true);
```
avoiding the default here as well

kadircet commented on clang-tools-extra/clangd/index/dex/Dex.h:
```diff
@@ -64,16 +65,18 @@ class Dex : public SymbolIndex {
   /// Builds an index from slabs. The index takes ownership of the slab.
-  static std::unique_ptr<SymbolIndex> build(SymbolSlab, RefSlab, RelationSlab);
+  static std::unique_ptr<SymbolIndex> build(SymbolSlab, RefSlab, RelationSlab,
+                                            bool SupportContainedRefs = true);
```
and here

kadircet commented on clang-tools-extra/clangd/tool/Check.cpp:
```diff
@@ -163,7 +163,7 @@ class Checker {
 
   Checker(llvm::StringRef File, const ClangdLSPServer::Options &Opts)
-      : File(File), Opts(Opts) {}
+      : File(File), Opts(Opts), Index(/*SupportContainedRefs=*/false) {}
```
why false here? i think it's better if we use whatever clangd-server's default is

HighCommander4 replied:
My original thinking was that `--check` does not perform any outgoing calls requests. But thinking more about it, a common use of `--check` is to reproduce crashes, so the more codepaths it executes (including the one behind this flag) the better in that regard.

kadircet reviewed (approved):
thanks, mostly LG! i think not leaving the default parameter values are important here to make sure we don't forget to pass the top-level config in any of the relevant places (both now, but also in the future). i know it probably comes with a bunch of changes to tests though :/ (i wish clangd was better at updating signatures in such cases!)

HighCommander4 commented:
Updated to remove the default arguments.

Thanks for the reviews!

kadircet commented on clang-tools-extra/clangd/index/remote/Index.proto:
```diff
@@ -131,3 +131,21 @@ message Relation {
 }
+
+message ContainedRefsRequest {
+  required string id = 1;
```
sorry I missed the `required` bits here and below. it's [discouraged](https://protobuf.dev/programming-guides/dos-donts/#add-required) and we're having troubles consuming this internally now.

there's no particular reason for these fields to be `required`, we're already taking care of error-handling during marshalling.

my colleague is going to revert this, I am happy to land it tomorrow with relevant changes. sorry for the second back and forth @HighCommander4  :/

HighCommander4 replied:
No worries, please go ahead

joaosaffran commented:
@HighCommander4 it appears that you forgot to update a [usage](https://github.com/llvm/llvm-project/blob/c3d15188cfe243900895a4f2c4f36b84e14928b7/clang-tools-extra/clangd/benchmarks/IndexBenchmark.cpp#L25-L33) of the changed loadIndex function.  This usage may be dead-coded due to the likely missing definition of a [BENCHMARK macro](https://github.com/llvm/llvm-project/blob/c3d15188cfe243900895a4f2c4f36b84e14928b7/clang-tools-extra/clangd/benchmarks/IndexBenchmark.cpp#L76-L98). Such usage broke one of HLSL build pipelines. Commenting, just for you to be aware of this issue when fixing it in the future.

HighCommander4 commented:
> relanded with benchmark fix in [61fe67a](https://github.com/llvm/llvm-project/commit/61fe67a4017375fd675f75652e857e837f77fa51)
> 
> changed required -> optional in [c7ef0ac](https://github.com/llvm/llvm-project/commit/c7ef0ac9fd28cb55b8c7c91a890b365cc688f9a9)

Thanks!

pidgeon777 commented:
Hello everyone, I'm seeking clarification regarding the implementation of Outgoing Calls support.

The LSP protocol currently allows to implement two display modes for visualizing outgoing function calls:

```c
#include <stdio.h>

void function1(void) {
    printf("Executing function 1\n");
}

void function2(void) {
    printf("Executing function 2\n");
}

void function3(void) {
    printf("Executing function 3\n");
}

int main(void) {
    printf("Starting main function\n");
    function1();
    function2();
    function3();
    return 0;
}
```

For example, when querying all functions called from `main`, results could be presented in two ways:

1. Display all the called function occurrences in `main`:
   - function1()
   - function2()
   - function2()
   - function3()

2. Display the definitions of the functions called in `main`:
   - function1()
   - function2()
   - function3()

According to one of the developers, this feature should provide all the necessary data to allow the LSP client to determine the presentation method (approach 1 or 2). I previously asked about this in another thread (which I cannot locate now), and I would appreciate confirmation that this will indeed be possible with the upcoming **clangd** changes.

HighCommander4 commented:
> I would appreciate confirmation that this will indeed be possible with the upcoming **clangd** changes.

Yes, the returned data is sufficient for clients to implement either of the above presentations.

pidgeon777 commented:
> > I would appreciate confirmation that this will indeed be possible with the upcoming **clangd** changes.
> 
> Yes, the returned data is sufficient for clients to implement either of the above presentations.

This is great news, thank you!

HighCommander4 commented:
> Any chance this is backportable to the 19.x series? Or is it too dependent on changes since then?
> 
> Or even better 18.x? (since a lot of downstream tools are dependent on 18.x and will be a while before they bump to 19)

If you're asking about the change being backported to appear in an official 19.x release, then I believe it would not be eligible for that; the release branch only accepts backports of regression and other serious bug fixes, not new features.

If you're asking whether the patch would apply cleanly to the 19.x release branch, for the purpose of applying it to some downstream repository that's based on llvm 19.x: while I haven't tried, I expect it should be possible to rebase the patch with fairly minimal changes. This is probably true of 18.x as well. If you try this and have a question about a specific conflict, I'm happy to provide suggestions.

pidgeon777 commented:
While I don't have direct experience with LLVM releases, I suppose the Outgoing Calls feature will be part of version 20?

HighCommander4 commented:
> While I don't have direct experience with LLVM releases, I suppose the Outgoing Calls feature will be part of version 20?

Yep

HighCommander4 commented:
> > @kadircet perhaps you might be able to pick up this review?
> > Or, in the absence of a full review, your opinion on the directional question in [this comment](https://reviews.llvm.org/D93829#4654786) would be appreciated as well:
> > > how would you feel about proceeding with the patch in its current state, with the memory usage increase brought down from 8.2% to 2.5% thanks to the combination of the simple lookup optimization + RefKind filtering, and leaving the "deep lookup optimization" to be explored in a future change?
> 
> I'd definitely prefer the one we discussed in the original review, but I don't think perfect needs to be enemy of the good, we can surely optimize data structures here going forward if needed.

I filed https://github.com/clangd/clangd/issues/2264 as a follow-up to track implementation of the "deep lookup optimization".

pidgeon777 commented:
> > While I don't have direct experience with LLVM releases, I suppose the Outgoing Calls feature will be part of version 20?
> 
> Yep

Does anyone know the estimated release date for version 20?

HighCommander4 commented:
> Does anyone know the estimated release date for version 20?

Based on https://llvm.org/docs/HowToReleaseLLVM.html#annual-release-schedule, approximately March 6.
